crates/tarball/src/lib.rs

```diff
@@ -62,59 +109,99 @@ fn decompress_gzip(gz_data: &[u8], unpacked_size: Option<usize>) -> Result<Vec<u
     Ok(decompressed)
 }
 
-#[instrument]
-fn verify_checksum(data: &[u8], integrity: &str) -> Result<(), TarballError> {
-    let validation = IntegrityChecker::new(Integrity::from_str(integrity)?).chain(data).result();
-
-    if validation.is_err() {
-        Err(TarballError::ChecksumMismatch)
-    } else {
-        Ok(())
-    }
+#[instrument(skip(data), fields(data_len = data.len()))]
+fn verify_checksum(data: &[u8], integrity: Integrity) -> Result<ssri::Algorithm, ssri::Error> {
+    integrity.pipe(IntegrityChecker::new).chain(data).result()
 }
 
-// #[instrument]
+#[instrument(skip(cache), fields(cache_len = cache.len()))]
 pub async fn download_tarball_to_store(
+    cache: &Cache,
     store_dir: &Path,
     package_integrity: &str,
     package_unpacked_size: Option<usize>,
     package_url: &str,
-) -> Result<HashMap<String, PathBuf>, TarballError> {
-    let http_client = Client::new();
-    let response = http_client.get(package_url).send().await?.bytes().await?;
-
-    let package_integrity = package_integrity.to_owned();
-    let store_dir = store_dir.to_owned();
-    tokio::task::spawn(async move {
-        verify_checksum(&response, &package_integrity)?;
-        let data = decompress_gzip(&response, package_unpacked_size).unwrap();
-        let mut archive = Archive::new(Cursor::new(data));
-        let cas_files = archive
-            .entries()
-            .unwrap()
+) -> Result<Arc<HashMap<String, PathBuf>>, TarballError> {
+    while let Some(cache_lock) = cache.get(package_url) {
+        tracing::info!(target: "pacquet::download", ?package_url, "Job taken");
+
+        match &*cache_lock.read().await {
+            CacheValue::Available(cas_paths) => {
+                tracing::info!(target: "pacquet::download", ?package_url, cas_paths_len = cas_paths.len(), "Cache hit");
+                return Ok(cas_paths.clone());
+            }
+            CacheValue::InProgress => {
+                tracing::info!(target: "pacquet::download", ?package_url, "Wait for cache");
+            }
+        }
+        drop(cache_lock);
+        sleep(Duration::from_millis(100)).await; // TODO: millis can be any small number, even 0, further testing is required to find the ideal number
```

Comment:
Is it OK to use sleep for this? Looks like a workaround.

Reply:
Why do we even need sleep in here? We need more comments on the code in general.